<a href='http://www.algebra.hr'> <img src='../algebra_logo_color_h.png' alt="Algebra" width="500" /></a>
___
# Pandas - Python Data Analysis Library

Znamo da Python ima nekoliko tipova podataka koji se odnose na kolekcije. To su liste, rječnici, n-terci. Rad s ovim tipovima podataka nije tako brz kao u jezicima baziranim na C-u ili u Fortranu. Zbog toga su nastale biblioteke koje imaju učinkovitost kao da su pisane u C-u ili Fortranu, a istovremeno imaju jednostavnost pisanja kôda kao u Pythonu.

Pandas je jedna od njih. To je Python biblioteka za analizu i manipulaciju podataka, strukturiranih u obliku  tablice. Pandas koristi NumPy biblioteku kao podlogu. NumPy ćemo upoznati kasnije, a njegova osnovna karakteristika je također rad s tabličnim tipom podataka koji može pohraniti samo podatke koji se zbrajaju i množe (najčešće brojeve). Pandas, s druge strane, također radi s tabličnim tipom podataka, ali za razliku od NumPy biblioteke, nema uvjet koji se tip podatka može pohraniti u tablici.

Pandas je jako korisna biblioteka i ovo su samo neke od situacija u kojima se koristi:<br>
- upravljanje podacima koji nedostaju u tablicama s brojevima (na njihova mjesta dodaje *NaN - Not a Number*)
- mogućnost promjene veličine objekta tako što se dodaju ili oduzimaju kolone iz *DataFrame* objekta
- fleksibilno, a opet učinkovito grupiranje podataka koje omogućava modifikacije kao što su podjela, dodavanje, kombiniranje...
- jednostavna konverzija Python i NumPy struktura podataka, bez obzira na način indeksacije izvornih objekata u *DateFrame* objekt
- mogućnost izdvajanja dijelova podataka u zasebne cjeline na osnovi naziva kolona
- promjena oblika i rotacija (eng. *pivoting*) tablica
- uvoz i izvoz podataka iz datoteka CSV, Excel, baze podataka.

... i još puno drugih.

## Instalacija i uporaba

Dolazi kao dio Anaconda distribucije, a Pandas paket moguće je instalirati i koristiti zasebno pomoću naredbe:

In [1]:
# pip install pandas

Više detalja o instalaciji Pandas paketa na adresi: https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html

Nakon instalacije, Pandas trebamo uključiti u skriptu, odnosno program:

In [2]:
import pandas as pd

Uobičajeno je da se Pandas dodaje tako da se preimenuje u "pd".

## Struktura podataka

Već smo naveli da se Pandas koristi za rad s podacima u obliku tablice. Nudi mogućnost da se učita cijela tablica i od nje napravi Python objekt s kojim možemo raditi svakojake manipulacije na podacima unutar tog objekta.

### *DataFrame*

Tablični oblik pohrane podataka zastupljen je u bazama podataka, tabličnim kalkulatorima (Excel, CSV). Ovakav, tablični tip podataka u Pandasu naziva se ***DataFrame*** i on predstavlja tablicu koja ima više redaka i stupaca. *DataFrame* ima retke i stupce pa kažemo da ima dvije dimenzije (os 0 - stupci i os 1 - redovi).

<img src='./01_table_dataframe.svg' alt="DataFrame" width="500" />
<center>Slika je preuzeta s:<br >
<a href='https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/01_table_oriented.html'> Pandas Documentation </a><center><br >

#### Kreiranje *DataFrame* objekta

Kreirajmo jedan *DataFrame* da nam bude malo jasnije.

Prvo ćemo uključiti Pandas biblioteku.

In [3]:
import pandas as pd

**NAPOMENA:** Mi ćemo u ovom slučaju kreirati jednu varijablu tipa *dict* i od nje ćemo kreirati tablicu. Vrijednost ključeva rječnika koristit će se kao nazivi stupaca, a kao vrijednost ćemo imati liste čiji će elementi biti dodijeljeni u redove svakog stupca. **Neki elementi lista su namjerno izbačeni, odnosno neke liste imaju manji broj članova!**

In [4]:
data = {
    'Kolona 1':['K1_Podatak 1', 'K1_Podatak 2', 'K1_Podatak 3', 'K1_Podatak 4', 'K1_Podatak 5'],
    'Kolona 2':['K2_Podatak 1', 'K2_Podatak 2', 'K2_Podatak 3', 'K2_Podatak 4', 'K2_Podatak 5'],
    'Kolona 3':['K3_Podatak 1', 'K3_Podatak 2', 'K3_Podatak 3', 'K3_Podatak 4', 'K2_Podatak 5'],
    'Kolona 4':['K4_Podatak 1', 'K4_Podatak 2', 'K4_Podatak 3', 'K4_Podatak 4', 'K4_Podatak 5']
}

Sada možemo kreirati jednu varijablu u koju ćemo pohraniti našu tablicu pretvorenu u Python objekt, a koju ćemo dalje nazivati *DataFrame*. Dodatna informacija: često ćete u primjerima na internetu naći da se ove varijable nazivaju generičkim nazivom *df* kao *DataFrame*.

In [5]:
# na internetu bi ova linija izgledala ovako:
# df = pd.DataFrame(data)
data_frame = pd.DataFrame(data)
data_frame

Kolona 1      Kolona 2      Kolona 3      Kolona 4
0  K1_Podatak 1  K2_Podatak 1  K3_Podatak 1  K4_Podatak 1
1  K1_Podatak 2  K2_Podatak 2  K3_Podatak 2  K4_Podatak 2
2  K1_Podatak 3  K2_Podatak 3  K3_Podatak 3  K4_Podatak 3
3  K1_Podatak 4  K2_Podatak 4  K3_Podatak 4  K4_Podatak 4
4  K1_Podatak 5  K2_Podatak 5  K2_Podatak 5  K4_Podatak 5

Obratite pažnju na prvu kolonu s brojevima od 0 do 4. To je kolona s indeksima redaka. Pandas je kolonu s indeksima sam kreirao. Kasnije ćemo se upoznati s mogućnostima prilagodbe ove opcije.

Isto tako, nismo morali koristiti print() funkciju za ispis, nego je bilo dovoljno samo navesti naziv *DataFrame* objekta i dobili smo formatirani ispis našeg rječnika u obliku tablice.

Vratit ćemo se na detaljne primjere rada s *DataFrame* objektom malo kasnije.

### Pandas *Series* / Serija

Drugi oblik podataka, za čiju obradu se koristi Pandas je *Series* ili serija kako ćemo je mi zvati. Serija predstavlja stupac u *DataFrame* tablici i zbog toga ima samo jednu dimenziju (os 0).

<img src='./01_table_series.svg' alt="DataFrame" heght="200" />
<center>Slika je preuzeta s:<br >
<a href='https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/01_table_oriented.html'> Pandas Documentation </a><center><br >


#### Kreiranje Serije

Rekli smo da je serija jednodimenzionalni Pandas objekt, jednostavnije, to je jedan stupac neke tablice ili *DataFrame* objekta. Seriju možemo kreirati na više načina pa krenimo s najčešćim, odnosno kreiranjem iz kolone prethodno kreiranog *DataFrame* objekta.

Uzet ćemo za primjer kolonu 3 i od nje kreirati seriju.

In [6]:
data_frame["Kolona 3"]

0    K3_Podatak 1
1    K3_Podatak 2
2    K3_Podatak 3
3    K3_Podatak 4
4    K2_Podatak 5
Name: Kolona 3, dtype: object

U liniji 1 gornje skripte, trebali biste prepoznati sintaksu za dohvat "vrijednosti" rječnika koja je pohranjena pod "ključem" navedenim u uglatim zagradama. *DataFrame* smo kreirali na osnovi rječnika, za dohvat jednog objekta *DataFramea* koristimo njegov naziv i ključ pod kojim je stupac pohranjen.

**Važno je zapamtiti: ovo je samo sintaksa za pristup stupcu *DataFrame* objekta, ali ovdje se NE radi o rječniku, već o objektu koji je Pandas *DataFrame* tip podatka!**

Primijetite još da serija nema naslov nego samo vrijednosti, odnosno na kraju, u zadnjem redu ispisa, naveden je naziv serije preuzet od naziva kolone *DataFramea*.

Kreirajmo novi Pandas serijski objekt koji nije vezan uz *DataFrame*.

In [7]:
serija = pd.Series([5, 8, 2, 6, 5, 4, 9], name='Serija')
serija

0    5
1    8
2    2
3    6
4    5
5    4
6    9
Name: Serija, dtype: int64

Kreirali smo novi Pandas *Series* objekt od liste brojeva te smo mu dali ime Oznaka. Listu brojeva smo mogli kreirati zasebno u jednu varijablu te onda tu varijablu predati kao argument u *Pandas.Series()* metodu.

Kako seriju nismo vezali uz kolonu nekog *DataFramea*, korisno je navesti i naziv serije pa smo za to koristili argument name. Taj naziv će se prikazati u ispisu serije. Osim naziva serije, uvijek dobijemo i podatak o tipu podataka koji su pohranjeni u seriji (*dtype*). U ovom našem slučaju to je *int64*, odnosno *Integer* ili cijeli brojevi.

**ZADATAK**<br>
Kreirajte nekoliko novih serija sa sljedećim tipovima podataka: *string*, decimalni brojevi, *Boolean* vrijednosti, Python liste.

In [8]:
# U ovu ćeliju upisište Vaše rješenje

Pandas serija uvijek kreira jednu kolonu, u kojoj su indeksi retka u kojemu se čuva određeni podatak. Predefinirana vrijednost su cijeli brojevi počevši od 0 (nula). Za indekse možemo koristiti i druge opcije, recimo slova.

**NAPOMENA: Broj članova liste indeksa MORA biti jednak broju članova liste koju smo predali kao *DATA* argument!**

In [9]:
serija_indeksi = pd.Series([5, 8, 2, 6], index=['a', 'b', 'c', 'd'])
serija_indeksi

a    5
b    8
c    2
d    6
dtype: int64

#### Pristup i izmjena članova serije

Kreirajmo Pandas seriju.

*Dodatna informacija: mogli smo koristiti varijablu iz prethodno kreirane ćelije ovog Jupyter Notebook dokumenta, ali ovaj put ćemo napraviti iznimku da primjer bude jasniji, a u seriju ćemo dodati još elemenata (uključujući i elemente bez vrijednosti).*

In [10]:
serija = pd.Series([5, 8, 2, 6, 5, 4, 9, 7, 1, 3, 5, 3, 4, 8, 9, 8, 2, 3, 5, 2], name='Oznaka')
serija

0     5
1     8
2     2
3     6
4     5
5     4
6     9
7     7
8     1
9     3
10    5
11    3
12    4
13    8
14    9
15    8
16    2
17    3
18    5
19    2
Name: Oznaka, dtype: int64

Pristupimo 5. elementu serije.

In [11]:
serija[5]

4

Sintaksa je ista kao za Python liste. Provjerimo, vrijedi li sintaksa za dohvat više elemenata istovremeno?

In [12]:
serija[:]

0     5
1     8
2     2
3     6
4     5
5     4
6     9
7     7
8     1
9     3
10    5
11    3
12    4
13    8
14    9
15    8
16    2
17    3
18    5
19    2
Name: Oznaka, dtype: int64

In [13]:
serija[3:8]

3    6
4    5
5    4
6    9
7    7
Name: Oznaka, dtype: int64

In [14]:
serija[3:8:3]

3    6
6    9
Name: Oznaka, dtype: int64

**Obratite pažnju na zadnji red u kojem je naveden naziv serije te tip podatka - i dalje se radi o istoj seriji, samo su prikazani pojedinačni elementi.**

Pokušajmo pohraniti izdvojene elemente u novu varijablu i provjerimo njezin tip i naziv.

In [15]:
podskup = serija[2:8:3]
podskup

2    2
5    4
Name: Oznaka, dtype: int64

Nema promjena, to je zato jer se i dalje radi o istom Pandas objektu. Kako ćemo onda kreirati novu seriju? Pokušajmo ovaj izdvojeni podskup podataka proslijediti kao *DATA* argument u *Pandas.Series()* metodu.

Prije toga, bilo bi dobro koristiti *help()* metodu, koja nam može pomoći ili dati neke smjernice.

In [ ]:
help(pd.Series)

Argument *"copy : bool, default False | Copy input data."* mogao bi biti zanimljiv zato jer smo sa svim dosadašnjim operacijama **uvijek čuvali reference na izvorni objekt.**

In [16]:
podskup = pd.Series(serija[2:8:3], name='Podskup', copy=True)
podskup

2    2
5    4
Name: Podskup, dtype: int64

Ako ispišemo varijablu "serija", vidjet ćemo da smo sada kreirali novi Pandas *Series* objekt.

In [17]:
serija

0     5
1     8
2     2
3     6
4     5
5     4
6     9
7     7
8     1
9     3
10    5
11    3
12    4
13    8
14    9
15    8
16    2
17    3
18    5
19    2
Name: Oznaka, dtype: int64

Čitanje, odnosno dohvat podataka i sintaksa su identične kao za Python liste. Provjerimo kako radi pridruživanje podataka, odnosno izmjena podataka pojedinačnih elemenata.

Za primjer ćemo koristiti novi Pandas *Series* objekt "podskup" i promijenit ćemo prvi element.

In [18]:
podskup[0]

KeyError: 0

Dogodio se problem. Opis greške je "ValueError: 0 is not in range". Očigledno nemamo indeks 0 (nula). Podskup smo dobili od elemenata iz serije "serija". Ispišimo ponovno članove serije "podskup".

In [19]:
podskup

2    2
5    4
Name: Podskup, dtype: int64

Prva kolona prikazuje brojeve 2 i 5 kao indekse nove Pandas serije naziva Podskup, kreirane od elemenata druge Pandas serije naziva Serija. Čuvanje indeksa može biti korisno, ali korisnije je ako za kolonu Index, koja se automatski kreira, znamo da uvijek počinje od 0. Ako se indeks kolona kreira automatski, onda se može jednostavno resetirati. Idemo to napraviti pomoću metode *reset_index()*.

Prvo ćemo dohvatiti listu indeksa i listu vrijednosti (rječnik - ključ/vrijednost) kako bismo provjerili njihove vrijednosti. Ovo nije nužno raditi prije resetiranja indeksa, nego je pokazano kako se to radi, a može biti korisno pri obradi podataka.

In [20]:
podskup.index

RangeIndex(start=2, stop=8, step=3)

In [21]:
podskup.values

array([2, 4])

In [22]:
podskup.reset_index()

index  Podskup
0      2        2
1      5        4

Provjerimo kako sada izgleda Pandas serija *podskup*.

In [23]:
podskup

2    2
5    4
Name: Podskup, dtype: int64

Očigledno smo napravili izmjenu, ali tako da smo kreirali novi objekt koji nismo nigdje pohranili pa smo ga nakon završetka izvođenja skripte izgubili (provjerite koji bi to bio tip novog objekta - metoda *type()*).

Naš cilj je napraviti "trajnu" izmjenu indeksa, ali na našem postojećem objektu. Zato u metodu *reset_index()* trebamo dodati dva argumenta:<br>
- *inplace=True* - napraviti to nad postojećim objektom
- *drop=True* - uz *inplace* obavezan argument, osigurava to da se postojeći indeksi odbace i kreiraju novi. 

Provjerimo je li pomoglo.

In [24]:
podskup.reset_index(inplace=True, drop=True)
podskup

0    2
1    4
Name: Podskup, dtype: int64

Sada je već puno bolje. Vratimo se na mijenjanje podataka u Pandas seriji. Promijenimo prvi element serije i ispišimo seriju.

In [25]:
podskup[0] = 79
podskup

0    79
1     4
Name: Podskup, dtype: int64

Probajmo promijeniti prvi element u neki podatak drugačijeg tipa, recimo u tekst.

In [26]:
podskup[0] = 'Prvi element'
podskup

0    Prvi element
1               4
Name: Podskup, dtype: object

Dakle, možemo dodavati i druge tipove podataka. Isto kao i u Python listi.

#### Filtriranje podataka

Zahvaljujući činjenici da je Pandas razvijen na NumPy biblioteci, dobili smo neke funkcionalnosti nad Pandas tipovima podataka. Probajmo filtrirati podatke u seriji tako da dohvatimo samo one veće od 6.

In [27]:
serija

0     5
1     8
2     2
3     6
4     5
5     4
6     9
7     7
8     1
9     3
10    5
11    3
12    4
13    8
14    9
15    8
16    2
17    3
18    5
19    2
Name: Oznaka, dtype: int64

In [28]:
serija[serija > 6]

1     8
6     9
7     7
13    8
14    9
15    8
Name: Oznaka, dtype: int64

Uzmimo, na primjer, da naša serija sadrži postotke uz neke podatke. Da bismo mogli računati s postotcima,  moramo ih pretvoriti u decimalni oblik 0,nn (5% = 0,05). Umjesto da koristimo *for* petlju u Pythonu, to možemo napraviti puno jednostavnije, tako da varijablu *serija* podijelimo sa 100.

In [29]:
serija / 100

0     0.05
1     0.08
2     0.02
3     0.06
4     0.05
5     0.04
6     0.09
7     0.07
8     0.01
9     0.03
10    0.05
11    0.03
12    0.04
13    0.08
14    0.09
15    0.08
16    0.02
17    0.03
18    0.05
19    0.02
Name: Oznaka, dtype: float64

Sada možemo koristiti vrijednosti podataka u seriji kao postotak.

Možda ste primijetili da serija ima dosta istih brojeva. Napravimo filter koji će nam prikazati vrijednosti elemenata samo jednom, bez obzira koliko se puta taj element ponavljao.

In [30]:
serija.unique()

array([5, 8, 2, 6, 4, 9, 7, 1, 3])

Dobro je znati koje elemente imamo jedinstvene u našoj seriji, ali bi bilo dobro imati i informaciju koliko puta se taj broj ponavlja. To nam omogućava metoda *value_counts()* dostupna nad Pandas *Series* objektom. 

Provjerimo.

In [31]:
serija.value_counts().sort_values()

1    1
6    1
7    1
4    2
9    2
2    3
3    3
8    3
5    4
Name: Oznaka, dtype: int64

Prvi stupac prikazuje vrijednost (jedinstvenu), a drugi stupac prikazuje koliko se puta pojavljuje ta vrijednost u seriji. Dodali smo i još jednu metodu *sort_values()* koja sortira vrijednosti serije, što se može zaključiti iz naziva metode.

Ponekad želimo provjeriti postoje li u našoj seriji neki elementi. Vidjet ćemo kasnije zašto nam je to korisno.

In [32]:
serija.isin([0, 5, 9])

0      True
1     False
2     False
3     False
4      True
5     False
6      True
7     False
8     False
9     False
10     True
11    False
12    False
13    False
14     True
15    False
16    False
17    False
18     True
19    False
Name: Oznaka, dtype: bool

Ponekad je korisno umjesto *Boolean* vrijednosti *True/False* dohvatiti vrijednosti elemenata koji zadovoljavaju uvjet te njihove indekse. 

Čitljivije je.

In [33]:
serija[serija.isin([0, 5, 9])]

0     5
4     5
6     9
10    5
14    9
18    5
Name: Oznaka, dtype: int64

To je dovoljno informacija o Pandas *Series* objektu. Vrijeme je za vježbu.

**Zadatak**<br>
Generirajte nasumičnih 100 brojeva te ih pohranite u Python listu. Od te liste kreirajte Pandas seriju te joj dajte nekakav naziv.
- Provjerite, koliko ima brojeva koji se ponavljaju te koliko se puta svaki od njih ponavlja?
- Koji je broj najveći, a koji najmanji?
- Modificirajte podatke u seriji tako da se svedu na postotke (neka budu u rasponu od 0 do 1).
- Promijenite vrijednosti najvećeg broja tako da je identična vrijednosti najmanjeg broja. Provjerite koji je sada najveći broj.

**Zadatak**<br>
Tekst za obradu:<br>
*Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.*

Obradite gore navedeni tekst tako da od njega kreirate listu riječi koje imaju sva mala slova i nemaju znakove interpunkcije. Od te liste kreirajte Pandas seriju te joj dajte nekakav naziv.
- Provjerite, koliko ima riječi koje se ponavljaju te koliko puta se svaka od njih ponavlja?
- Provjerite koja je najduža, a koja najkraća riječ.

### *DataFrame*

Vratimo se na *DataFrame* objekt i upoznajmo se s njegovim mogućnostima.

Na početku smo kreirali jedan *DataFrame* (dalje u tekstu DF) na osnovi rječnika u kojem su podaci bili bez pravog smisla. Sada ćemo napraviti isto dok ne dođemo do dijela kada ćemo DF kreirati na osnovi podataka iz neke .csv datoteke. Koristit ćemo podatke o proizvodima (naziv, cijena, kategorija, ocjena).

In [34]:
proizvodi_baza = {
    'Naziv': [
        'Trek', 'Kona', 'Giant', 
        'Bianchi', 'Cosmo Ride', 'Neon', 
        'Zeeclo', 'Atomic', 'Head', 
        'Elan', 'Salomon', 'Rossignol'
    ],
    'Cijena': [
        7699.00, 4699.00, 5999.00, 
        22499.00, 2599.00, 1999.00, 
        2799.00, 1499.00, 1359.00, 
        1499.00, 1699.00, 999.00
    ],
    'Kategorija': [
        'Bicikl', 'Bicikl', 'Bicikl', 
        'Bicikl', 'E-Romobil', 'E-Romobil',
        'E-Romobil', 'Skije', 'Skije', 
        'Skije', 'Skije', 'Skije'
    ],
    'Ocjena': [
        6.70, 6.50, 6.65, 
        7.20, 8.64, 8.61, 
        8.59, 7.99, 8.15,
        8.05, 7.91, 6.10
    ]
}

proizvodi_svi = pd.DataFrame(proizvodi_baza)
proizvodi_svi

Naziv   Cijena Kategorija  Ocjena
0         Trek   7699.0     Bicikl    6.70
1         Kona   4699.0     Bicikl    6.50
2        Giant   5999.0     Bicikl    6.65
3      Bianchi  22499.0     Bicikl    7.20
4   Cosmo Ride   2599.0  E-Romobil    8.64
5         Neon   1999.0  E-Romobil    8.61
6       Zeeclo   2799.0  E-Romobil    8.59
7       Atomic   1499.0      Skije    7.99
8         Head   1359.0      Skije    8.15
9         Elan   1499.0      Skije    8.05
10     Salomon   1699.0      Skije    7.91
11   Rossignol    999.0      Skije    6.10

Možda nam ne trebaju svi podaci, nego samo "naziv proizvoda" te "cijena". Kreirajmo DF samo od tih podataka.

In [35]:
proizvodi_dio = pd.DataFrame(proizvodi_baza, columns=['Naziv', 'Cijena'])
proizvodi_dio

Naziv   Cijena
0         Trek   7699.0
1         Kona   4699.0
2        Giant   5999.0
3      Bianchi  22499.0
4   Cosmo Ride   2599.0
5         Neon   1999.0
6       Zeeclo   2799.0
7       Atomic   1499.0
8         Head   1359.0
9         Elan   1499.0
10     Salomon   1699.0
11   Rossignol    999.0

Ovaj naš DF objekt je jednostavan, ali ćemo često raditi s puno većim tablicama preuzetim iz .csv datoteka i trebat ćemo se upoznati s tim podacima. Jedan od koraka je provjeriti nazive svih kolona.

In [38]:
proizvodi_svi.columns

Index(['Naziv', 'Cijena', 'Kategorija', 'Ocjena'], dtype='object')

Onda postoji mogućnost dohvata samo vrijednosti, bez naziva kolona.

In [39]:
proizvodi_svi.values

array([['Trek', 7699.0, 'Bicikl', 6.7],
       ['Kona', 4699.0, 'Bicikl', 6.5],
       ['Giant', 5999.0, 'Bicikl', 6.65],
       ['Bianchi', 22499.0, 'Bicikl', 7.2],
       ['Cosmo Ride', 2599.0, 'E-Romobil', 8.64],
       ['Neon', 1999.0, 'E-Romobil', 8.61],
       ['Zeeclo', 2799.0, 'E-Romobil', 8.59],
       ['Atomic', 1499.0, 'Skije', 7.99],
       ['Head', 1359.0, 'Skije', 8.15],
       ['Elan', 1499.0, 'Skije', 8.05],
       ['Salomon', 1699.0, 'Skije', 7.91],
       ['Rossignol', 999.0, 'Skije', 6.1]], dtype=object)

**NAPOMENA: Kasnije ćemo pojasniti tip rezultata "array(\[...\])" zato jer je vezan uz NumPy.**

Možemo li možda dohvatiti i indekse kao u Pandas *Series* objektu? Provjerimo.

In [41]:
proizvodi_svi.index

RangeIndex(start=0, stop=12, step=1)

Indeksi se kreću u rasponu od 0 (uključeno) do 12 (isključeno) i korak je 1. Dakle, od 0 do 11.

Sve kolone imaju naziv preuzet iz "ključa" rječnika od kojeg smo kreirali DF, dok ga samo generirana kolona s indeksima nema. Dodajmo i njoj naziv "Id".

In [42]:
proizvodi_svi.index.name = 'Id'
proizvodi_svi

Naziv   Cijena Kategorija  Ocjena
Id                                        
0         Trek   7699.0     Bicikl    6.70
1         Kona   4699.0     Bicikl    6.50
2        Giant   5999.0     Bicikl    6.65
3      Bianchi  22499.0     Bicikl    7.20
4   Cosmo Ride   2599.0  E-Romobil    8.64
5         Neon   1999.0  E-Romobil    8.61
6       Zeeclo   2799.0  E-Romobil    8.59
7       Atomic   1499.0      Skije    7.99
8         Head   1359.0      Skije    8.15
9         Elan   1499.0      Skije    8.05
10     Salomon   1699.0      Skije    7.91
11   Rossignol    999.0      Skije    6.10

Ako želimo izdvojiti samo jednu kolonu (Pandas seriju), to već znamo. Treba samo navesti naziv DF-a i u uglatim zagradama naziv kolone.

In [43]:
proizvodi_svi['Ocjena']

Id
0     6.70
1     6.50
2     6.65
3     7.20
4     8.64
5     8.61
6     8.59
7     7.99
8     8.15
9     8.05
10    7.91
11    6.10
Name: Ocjena, dtype: float64

Ovo je bilo lako, no što kada trebamo dohvatiti jedan ili tri reda za koje znamo njihove indekse? Možda Python sintaksa može pomoći. Provjerimo.

In [44]:
proizvodi_svi[2]

KeyError: 2

Ne pomaže. Ako si tablicu predočimo kao listu (redove) unutar liste (kolone), možda bismo mogli koristiti sintaksu za dvostruku listu **naziv_liste\[kolona\]\[red\]**. Provjerimo.

In [45]:
proizvodi_svi['Cijena'][3]

22499.0

Radi, ali dobivamo samo jedan podatak.

Sintaksa dohvata redaka, kao u Python listama, NE radi za DF, nego dobijemo samo jedan podatak (jednu ćeliju). Radi li možda sintaksa za *slice* liste? Provjerimo tako što ćemo dohvatiti zadnjih 6 elemenata.

In [47]:
print(proizvodi_svi[ : 6 : -1])

proizvodi_svi[ : 11-6 : -1]

        Naziv  Cijena Kategorija  Ocjena
Id                                      
11  Rossignol   999.0      Skije    6.10
10    Salomon  1699.0      Skije    7.91
9        Elan  1499.0      Skije    8.05
8        Head  1359.0      Skije    8.15
7      Atomic  1499.0      Skije    7.99


Naziv  Cijena Kategorija  Ocjena
Id                                      
11  Rossignol   999.0      Skije    6.10
10    Salomon  1699.0      Skije    7.91
9        Elan  1499.0      Skije    8.05
8        Head  1359.0      Skije    8.15
7      Atomic  1499.0      Skije    7.99
6      Zeeclo  2799.0  E-Romobil    8.59

Ovo radi na način da i dalje koristimo *slice*, da prvo navedemo prvi element i zadnji element pa korak. Ako za korak koristimo negativan predznak, onda se krećemo u suprotnom smjeru. Važno je primijetiti da moramo ispravno definirati raspon. Najbolji način je uzeti zadnji indeks i od njega oduzeti onoliko elemenata koliko želimo dohvatiti.

DF ima jedan atribut naziva *loc* (od *location*) kojem predamo *listu* s jednim ili više indeksa. Probajmo:

In [48]:
proizvodi_svi.loc[[2, 5, 6]] # Obratite pažnju da je sintaksa kao lista u listi

Naziv  Cijena Kategorija  Ocjena
Id                                   
2    Giant  5999.0     Bicikl    6.65
5     Neon  1999.0  E-Romobil    8.61
6   Zeeclo  2799.0  E-Romobil    8.59

#### Dodavanje i brisanje podataka

Dodajmo novu kolonu u naš DF. Nazvat ćemo je 'Količina na skladištu'. Ako smo kolone dodali na osnovi ključa rječnika, onda bismo se mogli poslužiti tom sintaksom. Probat ćemo nekoliko opcija generiranja vrijednosti da vidimo kako se ponaša dodavanje kolona.

In [49]:
proizvodi_svi['Količina na skladištu'] = 15
proizvodi_svi

Naziv   Cijena Kategorija  Ocjena  Količina na skladištu
Id                                                               
0         Trek   7699.0     Bicikl    6.70                     15
1         Kona   4699.0     Bicikl    6.50                     15
2        Giant   5999.0     Bicikl    6.65                     15
3      Bianchi  22499.0     Bicikl    7.20                     15
4   Cosmo Ride   2599.0  E-Romobil    8.64                     15
5         Neon   1999.0  E-Romobil    8.61                     15
6       Zeeclo   2799.0  E-Romobil    8.59                     15
7       Atomic   1499.0      Skije    7.99                     15
8         Head   1359.0      Skije    8.15                     15
9         Elan   1499.0      Skije    8.05                     15
10     Salomon   1699.0      Skije    7.91                     15
11   Rossignol    999.0      Skije    6.10                     15

Možda da dodamo listu vrijednosti u novu kolonu. Prvo ćemo probati s nekoliko brojeva da vidimo kako se ponaša pa ćemo onda probati dodatne opcije.

In [50]:
proizvodi_svi['Količina na skladištu'] = [156, 5648, 5454, 125, 12, 4]
proizvodi_svi

ValueError: Length of values (6) does not match length of index (12)

***ValueError: Length of values (6) does not match length of index (12)* greška govori da moramo dodati točno onoliko brojeva koliko imamo redaka.**

In [51]:
proizvodi_svi['Količina na skladištu'] = [156, 5648, 5454, 125, 12, 4, 156, 5648, 5454, 125, 12, 4]
proizvodi_svi

Naziv   Cijena Kategorija  Ocjena  Količina na skladištu
Id                                                               
0         Trek   7699.0     Bicikl    6.70                    156
1         Kona   4699.0     Bicikl    6.50                   5648
2        Giant   5999.0     Bicikl    6.65                   5454
3      Bianchi  22499.0     Bicikl    7.20                    125
4   Cosmo Ride   2599.0  E-Romobil    8.64                     12
5         Neon   1999.0  E-Romobil    8.61                      4
6       Zeeclo   2799.0  E-Romobil    8.59                    156
7       Atomic   1499.0      Skije    7.99                   5648
8         Head   1359.0      Skije    8.15                   5454
9         Elan   1499.0      Skije    8.05                    125
10     Salomon   1699.0      Skije    7.91                     12
11   Rossignol    999.0      Skije    6.10                      4

Probajmo s generiranjem nasumičnih brojeva za ovu novu kolonu na stari Python način, tako što ćemo, kada budemo radili s NumPy bibliotekom, koristiti generator nasumičnih brojeva ove biblioteke koji je jednako tako učinkovit i koristan.

In [53]:
import random as rand

vrijednosti = []
for _ in range(len(proizvodi_svi.index)):
    vrijednosti.append(rand.randrange(100, 9999))

proizvodi_svi['Količina na skladištu'] = vrijednosti
proizvodi_svi

Naziv   Cijena Kategorija  Ocjena  Količina na skladištu
Id                                                               
0         Trek   7699.0     Bicikl    6.70                    192
1         Kona   4699.0     Bicikl    6.50                   7179
2        Giant   5999.0     Bicikl    6.65                   4027
3      Bianchi  22499.0     Bicikl    7.20                   2077
4   Cosmo Ride   2599.0  E-Romobil    8.64                   8737
5         Neon   1999.0  E-Romobil    8.61                   3187
6       Zeeclo   2799.0  E-Romobil    8.59                   7371
7       Atomic   1499.0      Skije    7.99                   8171
8         Head   1359.0      Skije    8.15                   6673
9         Elan   1499.0      Skije    8.05                   2428
10     Salomon   1699.0      Skije    7.91                   7063
11   Rossignol    999.0      Skije    6.10                   2419

Za brisanje kolone, uključujući i sadržaj kolone, koristimo komandu *del*.

In [54]:
del proizvodi_svi['Količina na skladištu']
proizvodi_svi

Naziv   Cijena Kategorija  Ocjena
Id                                        
0         Trek   7699.0     Bicikl    6.70
1         Kona   4699.0     Bicikl    6.50
2        Giant   5999.0     Bicikl    6.65
3      Bianchi  22499.0     Bicikl    7.20
4   Cosmo Ride   2599.0  E-Romobil    8.64
5         Neon   1999.0  E-Romobil    8.61
6       Zeeclo   2799.0  E-Romobil    8.59
7       Atomic   1499.0      Skije    7.99
8         Head   1359.0      Skije    8.15
9         Elan   1499.0      Skije    8.05
10     Salomon   1699.0      Skije    7.91
11   Rossignol    999.0      Skije    6.10

Već smo koristili naredbu *isin()* na Pandas *Series*. Ista naredba vrijedi i nad Pandas *DataFrame* objektima. Provjerimo, imamo li proizvod naziva Neon i pojavljuje li se negdje broj 1499.00?

In [55]:
proizvodi_svi.isin(['Neon', 1499.00])

Naziv  Cijena  Kategorija  Ocjena
Id                                   
0   False   False       False   False
1   False   False       False   False
2   False   False       False   False
3   False   False       False   False
4   False   False       False   False
5    True   False       False   False
6   False   False       False   False
7   False    True       False   False
8   False   False       False   False
9   False    True       False   False
10  False   False       False   False
11  False   False       False   False

Ili čitljivije formatirano:

In [56]:
proizvodi_svi[proizvodi_svi.isin(['Neon', 1499.00])]

Naziv  Cijena Kategorija  Ocjena
Id                                 
0    NaN     NaN        NaN     NaN
1    NaN     NaN        NaN     NaN
2    NaN     NaN        NaN     NaN
3    NaN     NaN        NaN     NaN
4    NaN     NaN        NaN     NaN
5   Neon     NaN        NaN     NaN
6    NaN     NaN        NaN     NaN
7    NaN  1499.0        NaN     NaN
8    NaN     NaN        NaN     NaN
9    NaN  1499.0        NaN     NaN
10   NaN     NaN        NaN     NaN
11   NaN     NaN        NaN     NaN

#### Čitanje i pisanje podataka

Tijekom ovog seminara, podatke ćemo učitavati iz .csv datoteka. Međutim, Pandas omogućava učitavanje podataka i iz drugih izvora kao što su baze podataka, JSON datoteke i sl. Čitanje i pisanje je omogućeno preko metoda koje za čitanje koriste englesku riječ read te dodatak naziva izvora. Za .csv datoteke to je metoda *read_csv()*. Za pisanje se koriste metode koje počinju s engleskom riječi to (prema) i opet je u nastavku naziv odredišta. Za .csv datoteku, to bi bila metoda *to_csv()*.

|**Čitanje**|**Pisanje**|
|---|---|
|read_csv|to_csv|
|read_sql|to_sql|
|read_json|to_JSON|
|read_clipboard|to_clipboard|

Kao što smo rekli, mi ćemo se tijekom ovog seminara koristiti .csv datotekama tako da ćemo koristiti *read_csv()* metodu, ali ćemo za demonstraciju koristiti pisanje u druge formate.

Iskoristimo Python *help()* metodu za upoznavanje s dostupnim opcijama metode *read_csv()*.

In [ ]:
help(pd.read_csv)

**NAPOMENA: za daljnje primjere, vježbe i zadatke koristit ćemo .csv datoteke iz mape *Data*.**<br>
**Sve datoteke su preuzete s javno dostupnih izvora koje svaka zemlja objavljuje na stranicama određenih statističkih zavoda. Ove datoteke namijenjene su znanstvenim istraživanjima i podučavanju.
Postoje i neke datoteke za čije se korištenje treba platiti, ali se one ne mogu preuzeti prije nego se izvrši uplata.**

Krenut ćemo s datotekom *'cars.csv'*. Idemo je učitati u *DataFrame* i prikazati.

In [58]:
cars_csv = pd.read_csv('../Data/Vjezbe_Primjeri/cars.csv')
cars_csv

Car;MPG;Cylinders;Displacement;Horsepower;Weight;Acceleration;Model;Origin
0    STRING;DOUBLE;INT;DOUBLE;DOUBLE;DOUBLE;DOUBLE;...                        
1    Chevrolet Chevelle Malibu;18.0;8;307.0;130.0;3...                        
2    Buick Skylark 320;15.0;8;350.0;165.0;3693.;11....                        
3    Plymouth Satellite;18.0;8;318.0;150.0;3436.;11...                        
4    AMC Rebel SST;16.0;8;304.0;150.0;3433.;12.0;70;US                        
..                                                 ...                        
402  Ford Mustang GL;27.0;4;140.0;86.00;2790.;15.6;...                        
403  Volkswagen Pickup;44.0;4;97.00;52.00;2130.;24....                        
404  Dodge Rampage;32.0;4;135.0;84.00;2295.;11.6;82;US                        
405    Ford Ranger;28.0;4;120.0;79.00;2625.;18.6;82;US                        
406     Chevy S-10;31.0;4;119.0;82.00;2720.;19.4;82;US                        

[407 rows x 1 columns]

Obratite pažnju na zadnju liniju u kojoj piše "407 rows × 1 columns". Dobili smo samo jednu kolonu. To nije dobro. Ako pogledamo rezultat koji smo dobili preko *help* metode za *read_csv()*, vidjet ćemo da imamo na raspolaganju *sep* argument (eng. *separator ili delimiter*). CSV je skraćenica od "*Comma-Separated Values*", odnosno zarezom odvojenih vrijednosti. Dakle, primarni *delimiter* je zarez, a mi u našem primjeru imamo ; kao *delimiter*. 

Ponovimo čitanje iz CSV datoteke, ali ovaj put ćemo dodati i argument za *delimiter*.

In [59]:
cars_csv = pd.read_csv('../Data/Vjezbe_Primjeri/cars.csv', sep=';')
cars_csv

Car     MPG Cylinders Displacement Horsepower  \
0                       STRING  DOUBLE       INT       DOUBLE     DOUBLE   
1    Chevrolet Chevelle Malibu    18.0         8        307.0      130.0   
2            Buick Skylark 320    15.0         8        350.0      165.0   
3           Plymouth Satellite    18.0         8        318.0      150.0   
4                AMC Rebel SST    16.0         8        304.0      150.0   
..                         ...     ...       ...          ...        ...   
402            Ford Mustang GL    27.0         4        140.0      86.00   
403          Volkswagen Pickup    44.0         4        97.00      52.00   
404              Dodge Rampage    32.0         4        135.0      84.00   
405                Ford Ranger    28.0         4        120.0      79.00   
406                 Chevy S-10    31.0         4        119.0      82.00   

     Weight Acceleration Model  Origin  
0    DOUBLE       DOUBLE   INT     CAT  
1     3504.         12.0    70      US  
2     3693.         11.5    70      US  
3     3436.         11.0    70      US  
4     3433.         12.0    70      US  
..      ...          ...   ...     ...  
402   2790.         15.6    82      US  
403   2130.         24.6    82  Europe  
404   2295.         11.6    82      US  
405   2625.         18.6    82      US  
406   2720.         19.4    82      US  

[407 rows x 9 columns]

Sada je već puno bolje. I dalje imamo 407 redaka, ali sada smo dobili 9 kolona.

Provjerimo što ćemo dobiti ako koristimo svojstva: *columns, values i index.*

In [61]:
cars_csv = pd.read_csv('../Data/Vjezbe_Primjeri/cars.csv', sep=';')
cars_csv

Car     MPG Cylinders Displacement Horsepower  \
0                       STRING  DOUBLE       INT       DOUBLE     DOUBLE   
1    Chevrolet Chevelle Malibu    18.0         8        307.0      130.0   
2            Buick Skylark 320    15.0         8        350.0      165.0   
3           Plymouth Satellite    18.0         8        318.0      150.0   
4                AMC Rebel SST    16.0         8        304.0      150.0   
..                         ...     ...       ...          ...        ...   
402            Ford Mustang GL    27.0         4        140.0      86.00   
403          Volkswagen Pickup    44.0         4        97.00      52.00   
404              Dodge Rampage    32.0         4        135.0      84.00   
405                Ford Ranger    28.0         4        120.0      79.00   
406                 Chevy S-10    31.0         4        119.0      82.00   

     Weight Acceleration Model  Origin  
0    DOUBLE       DOUBLE   INT     CAT  
1     3504.         12.0    70      US  
2     3693.         11.5    70      US  
3     3436.         11.0    70      US  
4     3433.         12.0    70      US  
..      ...          ...   ...     ...  
402   2790.         15.6    82      US  
403   2130.         24.6    82  Europe  
404   2295.         11.6    82      US  
405   2625.         18.6    82      US  
406   2720.         19.4    82      US  

[407 rows x 9 columns]

In [62]:
cars_csv.index

RangeIndex(start=0, stop=407, step=1)

In [63]:
cars_csv.columns

Index(['Car', 'MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight',
       'Acceleration', 'Model', 'Origin'],
      dtype='object')

In [64]:
cars_csv.values

array([['STRING', 'DOUBLE', 'INT', ..., 'DOUBLE', 'INT', 'CAT'],
       ['Chevrolet Chevelle Malibu', '18.0', '8', ..., '12.0', '70',
        'US'],
       ['Buick Skylark 320', '15.0', '8', ..., '11.5', '70', 'US'],
       ...,
       ['Dodge Rampage', '32.0', '4', ..., '11.6', '82', 'US'],
       ['Ford Ranger', '28.0', '4', ..., '18.6', '82', 'US'],
       ['Chevy S-10', '31.0', '4', ..., '19.4', '82', 'US']], dtype=object)

Svojstva *columns* i *index* su nam korisna, dok values i nije toliko. Puno korisnije informacije možemo dobiti ako saznamo koji tip podatka se nalazi u svakoj koloni. Za to koristimo *DataFrame* svojstvo *dtypes*.

In [65]:
cars_csv.dtypes

Car             object
MPG             object
Cylinders       object
Displacement    object
Horsepower      object
Weight          object
Acceleration    object
Model           object
Origin          object
dtype: object

Tip podataka za sve kolone je "object", što najčešće predstavlja tekstualni tip podatka ili tip podatka nad kojim se ne mogu izvršavati matematičke operacije. Međutim, u ovom slučaju to nije točno. Postoje kolone, čak ih je većina, koje sadrže numeričke podatke.

Ako malo bolje pogledate prikaz *DataFramea*, vidjet ćete da je red s indeksom 0, red u kojem su navedeni tipovi podatka za svaku kolonu. Taj red nam nije potreban. Navedene podatke možemo dobiti pomoću metoda Pandas *DataFramea*. 

Izbrišimo ga.

In [66]:
cars_df = cars_csv.drop([0])
cars_df

Car   MPG Cylinders Displacement Horsepower Weight  \
1    Chevrolet Chevelle Malibu  18.0         8        307.0      130.0  3504.   
2            Buick Skylark 320  15.0         8        350.0      165.0  3693.   
3           Plymouth Satellite  18.0         8        318.0      150.0  3436.   
4                AMC Rebel SST  16.0         8        304.0      150.0  3433.   
5                  Ford Torino  17.0         8        302.0      140.0  3449.   
..                         ...   ...       ...          ...        ...    ...   
402            Ford Mustang GL  27.0         4        140.0      86.00  2790.   
403          Volkswagen Pickup  44.0         4        97.00      52.00  2130.   
404              Dodge Rampage  32.0         4        135.0      84.00  2295.   
405                Ford Ranger  28.0         4        120.0      79.00  2625.   
406                 Chevy S-10  31.0         4        119.0      82.00  2720.   

    Acceleration Model  Origin  
1           12.0    70      US  
2           11.5    70      US  
3           11.0    70      US  
4           12.0    70      US  
5           10.5    70      US  
..           ...   ...     ...  
402         15.6    82      US  
403         24.6    82  Europe  
404         11.6    82      US  
405         18.6    82      US  
406         19.4    82      US  

[406 rows x 9 columns]

Provjerimo ponovno tip podataka u ćelijama.

In [68]:
cars_df.dtypes

Car             object
MPG             object
Cylinders       object
Displacement    object
Horsepower      object
Weight          object
Acceleration    object
Model           object
Origin          object
dtype: object

I dalje je navedeno da se radi o 'object' tipu podataka. Pandas *DataFrame*, prilikom uvoza podataka, automatski prilagodi tip podatka, ali kako je u našem primjeru prvi podatak bio tekstualnog tipa (redak koji smo izbrisali), onda je cijela kolona ostavljena u tekstualnom tipu. U našem primjeru postoje kolone koje zaista imaju tekstualne podatke (naziv automobila u koloni *Car*) pa nije moguće koristiti naredbu *astype()* nad cijelim DF-om, već samo na pojedinačnim kolonama.

Za konverziju tipa podataka, koristi se metoda *astype()* koja kao argument prihvaća naziv tipa podataka. Metoda vraća konvertirane podatke na način da ih je potrebno pohraniti.

*Informativno: pokušajte pokrenuti **cars_df.astype(float).***

In [69]:
cars_df['MPG'] = cars_df['MPG'].astype('float')
cars_df

Car   MPG Cylinders Displacement Horsepower Weight  \
1    Chevrolet Chevelle Malibu  18.0         8        307.0      130.0  3504.   
2            Buick Skylark 320  15.0         8        350.0      165.0  3693.   
3           Plymouth Satellite  18.0         8        318.0      150.0  3436.   
4                AMC Rebel SST  16.0         8        304.0      150.0  3433.   
5                  Ford Torino  17.0         8        302.0      140.0  3449.   
..                         ...   ...       ...          ...        ...    ...   
402            Ford Mustang GL  27.0         4        140.0      86.00  2790.   
403          Volkswagen Pickup  44.0         4        97.00      52.00  2130.   
404              Dodge Rampage  32.0         4        135.0      84.00  2295.   
405                Ford Ranger  28.0         4        120.0      79.00  2625.   
406                 Chevy S-10  31.0         4        119.0      82.00  2720.   

    Acceleration Model  Origin  
1           12.0    70      US  
2           11.5    70      US  
3           11.0    70      US  
4           12.0    70      US  
5           10.5    70      US  
..           ...   ...     ...  
402         15.6    82      US  
403         24.6    82  Europe  
404         11.6    82      US  
405         18.6    82      US  
406         19.4    82      US  

[406 rows x 9 columns]

Provjerimo još jednom tipove podataka.

In [70]:
cars_df.dtypes

Car              object
MPG             float64
Cylinders        object
Displacement     object
Horsepower       object
Weight           object
Acceleration     object
Model            object
Origin           object
dtype: object

Sada smo dobili kolonu koja čuva podatke o potrošnji (milja po galonu).

**Zadatak**<br>
**Napravite konverziju preostalih kolona koje čuvaju numeričke podatke.**

In [71]:
cars_df['Cylinders'] = cars_df['Cylinders'].astype('int64')
cars_df['Displacement'] = cars_df['Displacement'].astype('float')
cars_df['Horsepower'] = cars_df['Horsepower'].astype('float')
cars_df['Weight'] = cars_df['Weight'].astype('float')
cars_df['Acceleration'] = cars_df['Acceleration'].astype('float')
cars_df['Model'] = cars_df['Model'].astype('int64')

In [72]:
cars_df.dtypes

Car              object
MPG             float64
Cylinders         int64
Displacement    float64
Horsepower      float64
Weight          float64
Acceleration    float64
Model             int64
Origin           object
dtype: object

In [73]:
cars_df

Car   MPG  Cylinders  Displacement  Horsepower  \
1    Chevrolet Chevelle Malibu  18.0          8         307.0       130.0   
2            Buick Skylark 320  15.0          8         350.0       165.0   
3           Plymouth Satellite  18.0          8         318.0       150.0   
4                AMC Rebel SST  16.0          8         304.0       150.0   
5                  Ford Torino  17.0          8         302.0       140.0   
..                         ...   ...        ...           ...         ...   
402            Ford Mustang GL  27.0          4         140.0        86.0   
403          Volkswagen Pickup  44.0          4          97.0        52.0   
404              Dodge Rampage  32.0          4         135.0        84.0   
405                Ford Ranger  28.0          4         120.0        79.0   
406                 Chevy S-10  31.0          4         119.0        82.0   

     Weight  Acceleration  Model  Origin  
1    3504.0          12.0     70      US  
2    3693.0          11.5     70      US  
3    3436.0          11.0     70      US  
4    3433.0          12.0     70      US  
5    3449.0          10.5     70      US  
..      ...           ...    ...     ...  
402  2790.0          15.6     82      US  
403  2130.0          24.6     82  Europe  
404  2295.0          11.6     82      US  
405  2625.0          18.6     82      US  
406  2720.0          19.4     82      US  

[406 rows x 9 columns]

Svo ovo smo mogli napraviti puno jednostavnije i brže. Gornji primjer je naveden zato da bismo pokazali kako se mogu raditi konverzije pojedinačnih kolona.

Vratimo se ispravnom uvozu podataka. Sada kada znamo da nam prvi red radi poteškoće, onda ćemo ga odmah tijekom uvoza izostaviti tako što ćemo koristiti argument *skiprows* koji predstavlja listu indeksa redova koje želimo izostaviti.

In [74]:
cars_csv = pd.read_csv('../Data/Vjezbe_Primjeri/cars.csv', sep=';', skiprows=[0])
cars_csv

STRING  DOUBLE  INT  DOUBLE.1  DOUBLE.2  DOUBLE.3  \
0    Chevrolet Chevelle Malibu    18.0    8     307.0     130.0    3504.0   
1            Buick Skylark 320    15.0    8     350.0     165.0    3693.0   
2           Plymouth Satellite    18.0    8     318.0     150.0    3436.0   
3                AMC Rebel SST    16.0    8     304.0     150.0    3433.0   
4                  Ford Torino    17.0    8     302.0     140.0    3449.0   
..                         ...     ...  ...       ...       ...       ...   
401            Ford Mustang GL    27.0    4     140.0      86.0    2790.0   
402          Volkswagen Pickup    44.0    4      97.0      52.0    2130.0   
403              Dodge Rampage    32.0    4     135.0      84.0    2295.0   
404                Ford Ranger    28.0    4     120.0      79.0    2625.0   
405                 Chevy S-10    31.0    4     119.0      82.0    2720.0   

     DOUBLE.4  INT.1     CAT  
0        12.0     70      US  
1        11.5     70      US  
2        11.0     70      US  
3        12.0     70      US  
4        10.5     70      US  
..        ...    ...     ...  
401      15.6     82      US  
402      24.6     82  Europe  
403      11.6     82      US  
404      18.6     82      US  
405      19.4     82      US  

[406 rows x 9 columns]

In [75]:
cars_df.dtypes

Car              object
MPG             float64
Cylinders         int64
Displacement    float64
Horsepower      float64
Weight          float64
Acceleration    float64
Model             int64
Origin           object
dtype: object

Dobili smo ispravno uvezene tipove podataka u jednom redu. Ovo je jako važno kada imamo velike datoteke koje imaju jako puno redova, ali i kolona, zato je prolazak kolonu po kolonu kako bi se promijenio tip podatka zamoran, neučinkovit i podliježe greškama.

*Dodatna informacija: postoje još dodatni argumenti koje ćemo upoznati kasnije ili čiju primjenu možete sami istraživati.*

Sada kada smo prilagodili tip podatka u kolonama na način da nama odgovara. Nastavimo dalje s upoznavanjem s našim podacima te njihovom normalizacijom.

Provjerimo imamo li prazne ćelije. To su ćelije koje ne čuvaju nikakve podatke. Za to koristimo svojstvo *empty*. Ovo svojstvo vraća True ako nađe praznu ćeliju, odnosno *False* ako nema praznih ćelija.

In [76]:
cars_df.empty

False

U našem primjeru imamo "samo" 406 redaka. Vidjet ćete iz kasnijih primjera da ovakve tablice imaju i po desetak tisuća redaka. Nakon uvoza podataka, koristili smo integrirani tablični prikaz DF-a koji prikazuje prvih 5 redova pa onda jedan red s po tri točkice, zadnjih pet redova te na kraju broj redova i kolona. Ponekad će i ovakav prikaz biti nejasan pa se ovi dijelovi prikaza mogu prikazati zasebno.

- *head()* - prikazuje prvih pet redova, a ako kao argument dajemo cijeli broj, vratit će nam toliko redova
- *tail()* - prikazuje zadnjih pet redova, a ako kao argument dajemo cijeli broj, vratit će nam toliko redova
- *shape* - svojstvo koje vraća broj redaka i broj kolona u obliku n-terca (broj_redaka, broj_kolona)

In [77]:
cars_df.head()

Car   MPG  Cylinders  Displacement  Horsepower  \
1  Chevrolet Chevelle Malibu  18.0          8         307.0       130.0   
2          Buick Skylark 320  15.0          8         350.0       165.0   
3         Plymouth Satellite  18.0          8         318.0       150.0   
4              AMC Rebel SST  16.0          8         304.0       150.0   
5                Ford Torino  17.0          8         302.0       140.0   

   Weight  Acceleration  Model Origin  
1  3504.0          12.0     70     US  
2  3693.0          11.5     70     US  
3  3436.0          11.0     70     US  
4  3433.0          12.0     70     US  
5  3449.0          10.5     70     US

In [78]:
cars_df.tail()

Car   MPG  Cylinders  Displacement  Horsepower  Weight  \
402    Ford Mustang GL  27.0          4         140.0        86.0  2790.0   
403  Volkswagen Pickup  44.0          4          97.0        52.0  2130.0   
404      Dodge Rampage  32.0          4         135.0        84.0  2295.0   
405        Ford Ranger  28.0          4         120.0        79.0  2625.0   
406         Chevy S-10  31.0          4         119.0        82.0  2720.0   

     Acceleration  Model  Origin  
402          15.6     82      US  
403          24.6     82  Europe  
404          11.6     82      US  
405          18.6     82      US  
406          19.4     82      US

In [79]:
cars_df.shape

(406, 9)

**VJEŽBA Dohvatite 3., 8. i 15 prvih, odnosno zadnjih redaka.**

Upoznajmo se s još jednom metodom za upoznavanje s podacima, odnosno cijelim DF-om. To je metoda *info()*.

In [80]:
cars_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 406 entries, 1 to 406
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Car           406 non-null    object 
 1   MPG           406 non-null    float64
 2   Cylinders     406 non-null    int64  
 3   Displacement  406 non-null    float64
 4   Horsepower    406 non-null    float64
 5   Weight        406 non-null    float64
 6   Acceleration  406 non-null    float64
 7   Model         406 non-null    int64  
 8   Origin        406 non-null    object 
dtypes: float64(5), int64(2), object(2)
memory usage: 31.7+ KB


Uvoz podatka u Pandas je jako opširno područje, no zajedno smo prošli najvažnije dijelove. Uvoz podataka iz drugih izvora, kao što su JSON, Excel i baze podataka (SQLite3 i sl.) nismo radili, ali se puno ne razlikuju. Za rad s bazama podataka UVIJEK trebate imati jedan objekt koji predstavlja konekciju prema bazi, taj ćete objekt trebati i za Pandas pa vodite računa da radite na način jedna tablica, jedan DF.

Pisanje u datoteke je također jednostavno. Koristite metodu *to_csv()* kojoj kao argument predate naziv datoteke i to je dovoljno. Naravno, možete koristiti i neke dodatne argumente za formatiranje, izbacivanje nekih podataka i slično.

**ZADACI**

Uvezite, normalizirajte te dokumentirajte sljedeće skupove podataka iz mape Data\Vjezbe_Primjeri:<br>
- *baby_names_France_from_nat1900-2017.tsv* - .tsv ekstenzija je isto kao i .csv, samo što je *tab delimiter*
- *e-cars.csv*
- *mall_customers.csv*
- *titanic_data.csv*
- *airbnb_new_york_city_reviews.csv*
- *airbnb_new_york_city_listings.csv*

*Dodatna informacija:  dokumentacija skupa podataka predstavlja opis toga kakav je skup bio u izvornom obliku (kakve podatke je imao, koliko redaka, kolona, praznih ćelija, tipova podatka ...) te kako skup izgleda nakon normalizacije. Ovo je važan korak jer ćete ponekad trebati samo pripremiti podatke, dok će njihovu obradu raditi netko drugi.*